### PR TITLE
run pyupgrade on io

### PR DESCRIPTION
**Pull request by nstarman**
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

I ran ``pyupgrade —py38-plus`` on ``astropy/io``.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/io/fits/tests/test_hdulist.py [hunk 1]
```diff
@@ -604,7 +604,7 @@ def test_open_file_with_bad_header_padding(self):
         padding_len = 2880 - padding_start
         with open(self.temp('temp.fits'), 'r+b') as f:
             f.seek(padding_start)
-            f.write('\0'.encode('ascii') * padding_len)
+            f.write(b'\0' * padding_len)
```
Comment (mhvk) [inline astropy/io/fits/tests/test_hdulist.py:607]: Quite impressive pattern recognition here!

**Review by mhvk — APPROVED** [review]
Looks good to me again.

**Review by saimn — APPROVED** [review]

File: astropy/io/ascii/mrt.py [hunk 2]
```diff
@@ -340,15 +341,14 @@ def write_byte_by_byte(self):
...
                         if col.min == col.max:
-                            lim_vals = "[{0}]".format(col.min)
+                            lim_vals = f"[{col.min}]"
                         else:
-                            lim_vals = "[{0}/{1}]".format(col.min, col.max)
+                            lim_vals = f"[{col.min}/{col.max}]"
                 elif col.fortran_format[0] in ('E', 'F'):
-                    lim_vals = "[{0}/{1}]".format(math.floor(col.min * 100) / 100.,
-                                                  math.ceil(col.max * 100) / 100.)
+                    lim_vals = f"[{floor(col.min * 100) / 100.}/{ceil(col.max * 100) / 100.}]"
```
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/io/ascii/mrt.py:348]: That one was actually me 😆.

[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/io/ascii/mrt.py:348]: I thought it would fit on one line and it did.

**Review by dhomeier — COMMENTED** [review]
I am afraid I have missed too much of the background of this change to form an opinion yet, but since others have approved already I am not objecting.

File: astropy/io/ascii/mrt.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/io/ascii/mrt.py:348]: Ah, that makes more sense. I don't mind the change, but couldn't imagine that pattern happening automatically!

[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/io/ascii/mrt.py:348]: Note that it is part of the string here, not of the expression!

[the remark this answers is not in the document]
Comment (dhomeier) [reply astropy/io/ascii/mrt.py:348]: Right, I missed that it's just replacing `{0}/{1}`!

File: astropy/io/ascii/fixedwidth.py [hunk 3]
```diff
@@ -132,7 +132,7 @@ def get_cols(self, lines):
                 # slice col_ends but expects inclusive col_ends on input (for
                 # more intuitive user interface).
                 line = self.get_line(lines, position_line)
-                if len(set(line) - set([self.splitter.delimiter, ' '])) != 1:
+                if len(set(line) - {self.splitter.delimiter, ' '}) != 1:
```
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/io/ascii/fixedwidth.py:135]: One could write this as `len(set(line).difference(self.splitter.delimiter, ' ')) != 1` to preserve clarity while still allowing `pyupgrade` to run automatically.

[the remark this answers is not in the document]
Comment (dhomeier) [reply astropy/io/ascii/fixedwidth.py:135]: Could be a smart workaround, although I am probably not object-oriented enough to find that clearer than arithmetic operators. 😆

[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/io/ascii/fixedwidth.py:135]: My suggestion would be to just keep the reformat here; in most other places, it is a real improvement.

File: astropy/io/ascii/tests/test_c_reader.py [hunk 4]
```diff
@@ -438,8 +438,8 @@ def test_quoted_fields(parallel, read_basic):
     ('data_start', -1),  # data_start negative
     ('quotechar', '##'),  # multi-char quote signifier
     ('header_start', -1),  # negative header_start
-    ('converters', dict((i + 1, ascii.convert_numpy(np.uint))
-                        for i in range(3))),  # passing converters
+    ('converters', {i + 1: ascii.convert_numpy(np.uint)
+                    for i in range(3)}),  # passing converters
```
Comment (dhomeier) [inline astropy/io/ascii/tests/test_c_reader.py:442]:
```suggestion
    ('converters', {i + 1: ascii.convert_numpy(np.uint) for i in range(3)}),  # passing converters
```

File: astropy/io/ascii/fixedwidth.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/io/ascii/fixedwidth.py:135]: pyupgrade is opinionated and not very configurable. As the creator writes repeatedly (and with an unnecessarily sharp tone) it's take-it-or-leave-it. Everything pyupgrade does should be the same speed or faster, as in set construction. Making sure a line keeps a consistent style will be up to us. I can modify this line to be ``len(set(line).difference(self.splitter.delimiter, ' ')) != 1`` , but personally I find ``len(set(line) - {self.splitter.delimiter, ' '})`` eager to read.

**Pull request merged**